File: cad/kpis.py
"""KPIs centralizados del proyecto.

Mirror en `web/js/kpis.js` — modificar ambos cuando cambies una fórmula.
Test de consistencia en `tests/test_kpis_consistency.py` (TODO).
"""
from __future__ import annotations

from dataclasses import dataclass
from math import hypot

from parameters import Params, Prices, load_params, load_prices


@dataclass(frozen=True)
class SteelLengths:
    """Longitudes totales por categoría, sin desperdicio (en metros)."""
    columns: float
    main_beams: float
    secondary_joists: float
    bracing: float
    rafters: float
    tie_beams: float
    ridge: float
    purlins: float

    @property
    def total(self) -> float:
        return (self.columns + self.main_beams + self.secondary_joists
                + self.bracing + self.rafters + self.tie_beams
                + self.ridge + self.purlins)


def steel_lengths(p: Params) -> SteelLengths:
    """Longitudes de acero antes de desperdicio."""
    width = p.width_m
    depth = p.depth_m
    cols = int(p.raw["platform"]["column_grid_cols"])
    rows = int(p.raw["platform"]["column_grid_rows"])

    columns = p.total_column_length_m
    main_beams = rows * width + cols * depth
    secondary_spacing = float(p.raw["platform_structure"]["joist_spacing_m"])
    joist_count = max(2, int(round(depth / secondary_spacing)) + 1)
    secondary_joists = joist_count * width
    bracing = 24.0  # estimación X-bracing entre postes (mejorable)

    n_portals = int(p.raw["aframe"]["portal_count"])
    rafter_len = p.rafter_length_m
    enclosed = p.enclosed_depth_m
    rafters = n_portals * 2 * rafter_len
    tie_beams = n_portals * width
    ridge = enclosed + 0.4  # con overhang
    purlin_rows = int(p.raw["aframe"]["purlin_rows_per_side"])
    purlins = purlin_rows * 2 * (enclosed + 0.4)

    return SteelLengths(
        columns=columns,
        main_beams=main_beams,
        secondary_joists=secondary_joists,
        bracing=bracing,
        rafters=rafters,
        tie_beams=tie_beams,
        ridge=ridge,
        purlins=purlins,
    )


def steel_weight_kg(p: Params, lengths: SteelLengths) -> float:
    return (
        lengths.columns          * p.profile(p.raw["columns"]["profile"]).kg_per_m
        + lengths.main_beams       * p.profile(p.raw["platform_structure"]["main_beam_profile"]).kg_per_m
        + lengths.secondary_joists * p.profile(p.raw["platform_structure"]["secondary_joist_profile"]).kg_per_m
        + lengths.bracing          * p.profile(p.raw["platform_structure"]["bracing_profile"]).kg_per_m
        + lengths.rafters          * p.profile(p.raw["aframe"]["rafter_profile"]).kg_per_m
        + lengths.tie_beams        * p.profile(p.raw["aframe"]["tie_beam_profile"]).kg_per_m
        + lengths.ridge            * p.profile(p.raw["aframe"]["ridge_profile"]).kg_per_m
        + lengths.purlins          * p.profile(p.raw["aframe"]["purlin_profile"]).kg_per_m
    )


def fabricated_steel_kg(raw_kg: float, fabrication_uplift: float = 1.125) -> float:
    """Acero crudo + 12.5% (promedio del rango 10–15%) por placas, soldadura,
    pernos y margen de fabricación."""
    return raw_kg * fabrication_uplift


@dataclass(frozen=True)
class AreaKPIs:
    enclosed_m2: float
    terrace_m2: float
    platform_m2: float
    roof_m2: float


def areas(p: Params) -> AreaKPIs:
    enclosed = p.width_m * p.enclosed_depth_m
    terrace = p.width_m * p.terrace_depth_m
    platform = enclosed + terrace
    overhang = float(p.raw["envelope"]["roof_overhang_m"])
    roof = 2 * p.rafter_length_m * (p.enclosed_depth_m + 2 * overhang)
    return AreaKPIs(enclosed_m2=enclosed, terrace_m2=terrace,
                    platform_m2=platform, roof_m2=roof)


@dataclass(frozen=True)
class CostKPIs:
    direct_cop: float
    indirect_cop: float
    contingency_cop: float
    pre_iva_cop: float
    iva_cop: float
    total_cop: float
    total_usd: float
    per_m2_enclosed_cop: float


def costs(p: Params, pr: Prices) -> CostKPIs:
    """Costo total estimado usando precios mid de prices.toml.

    Implementación M0.3.2 — simplificada. Itera el BOM completo en M0.3.3.
    Para este snapshot inicial uso el total mid del PRESUPUESTO.md manual
    (≈ $218.3 M directos) como anchor y aplicaré indirectos/contingencia/IVA
    desde prices.toml. Los generators de M0.3.3 producirán el cálculo
    detallado partida por partida.
    """
    # Anchor temporal — reemplazar en M0.3.3 con suma detallada del BOM.
    direct_cop = 218_322_300

    indirect = direct_cop * pr.indirect_pct
    contingency = direct_cop * pr.contingency_pct
    pre_iva = direct_cop + indirect + contingency
    iva = pre_iva * pr.iva_pct
    total = pre_iva + iva
    a = areas(p)

    return CostKPIs(
        direct_cop=direct_cop,
        indirect_cop=indirect,
        contingency_cop=contingency,
        pre_iva_cop=pre_iva,
        iva_cop=iva,
        total_cop=total,
        total_usd=total / pr.usd_cop,
        per_m2_enclosed_cop=total / a.enclosed_m2,
    )


@dataclass(frozen=True)
class AllKPIs:
    steel_lengths: SteelLengths
    steel_raw_kg: float
    steel_fabricated_kg: float
    anchors: int
    areas: AreaKPIs
    cost: CostKPIs


def all_kpis(p: Params, pr: Prices) -> AllKPIs:
    sl = steel_lengths(p)
    raw_kg = steel_weight_kg(p, sl)
    fab_kg = fabricated_steel_kg(raw_kg)
    return AllKPIs(
        steel_lengths=sl,
        steel_raw_kg=raw_kg,
        steel_fabricated_kg=fab_kg,
        anchors=p.anchor_count,
        areas=areas(p),
        cost=costs(p, pr),
    )


if __name__ == "__main__":
    p = load_params()
    pr = load_prices()
    k = all_kpis(p, pr)

    print("== Acero ==")
    print(f"  Longitud cruda total:    {k.steel_lengths.total:>8.1f} m")
    print(f"    columnas:              {k.steel_lengths.columns:>8.1f} m")
    print(f"    vigas principales:     {k.steel_lengths.main_beams:>8.1f} m")
    print(f"    viguetas:              {k.steel_lengths.secondary_joists:>8.1f} m")
    print(f"    arriostramiento:       {k.steel_lengths.bracing:>8.1f} m")
    print(f"    pares A-frame:         {k.steel_lengths.rafters:>8.1f} m")
    print(f"    tirantes:              {k.steel_lengths.tie_beams:>8.1f} m")
    print(f"    cumbrera:              {k.steel_lengths.ridge:>8.1f} m")
    print(f"    correas:               {k.steel_lengths.purlins:>8.1f} m")
    print(f"  Peso crudo:              {k.steel_raw_kg:>8.0f} kg")
    print(f"  Peso fabricado (×1.125): {k.steel_fabricated_kg:>8.0f} kg")

    print("\n== Áreas ==")
    print(f"  Cabaña cerrada:      {k.areas.enclosed_m2:>6.1f} m²")
    print(f"  Terraza:             {k.areas.terrace_m2:>6.1f} m²")
    print(f"  Plataforma total:    {k.areas.platform_m2:>6.1f} m²")
    print(f"  Cubierta:            {k.areas.roof_m2:>6.1f} m²")

    print(f"\n== Anclajes ==  {k.anchors}")

    print("\n== Costo ==")
    print(f"  Directo:        $ {k.cost.direct_cop:>14,.0f} COP")
    print(f"  Indirecto:      $ {k.cost.indirect_cop:>14,.0f} COP")
    print(f"  Contingencia:   $ {k.cost.contingency_cop:>14,.0f} COP")
    print(f"  Subtotal:       $ {k.cost.pre_iva_cop:>14,.0f} COP")
    print(f"  IVA 19%:        $ {k.cost.iva_cop:>14,.0f} COP")
    print(f"  TOTAL:          $ {k.cost.total_cop:>14,.0f} COP  (≈ USD {k.cost.total_usd:,.0f})")
    print(f"  Por m² cabaña:  $ {k.cost.per_m2_enclosed_cop:>14,.0f} COP")

File: web/js/kpis.js
// kpis.js — Cálculo de KPIs en el browser.
// Mirror de cad/kpis.py. Mantener consistencia: si cambias una fórmula aquí,
// cámbiala allá y corre el test de consistencia.

const FAB_UPLIFT = 1.125;

function profile(params, name) {
  const p = params.profiles?.[name];
  if (!p) throw new Error(`perfil desconocido: ${name}`);
  return p;
}

function rafterLengthM(params) {
  const halfSpan = params.platform.width_m / 2;
  const apex = params.aframe.apex_height_m;
  return Math.hypot(halfSpan, apex);
}

function columnCount(params) {
  return params.platform.column_grid_cols * params.platform.column_grid_rows;
}

function totalColumnLengthM(params) {
  return params.columns.heights_m.flat().reduce((a, b) => a + b, 0);
}

function steelLengths(params) {
  const w = params.platform.width_m;
  const d = params.platform.depth_m;
  const cols = params.platform.column_grid_cols;
  const rows = params.platform.column_grid_rows;

  const columns = totalColumnLengthM(params);
  const main_beams = rows * w + cols * d;
  const joistSpacing = params.platform_structure.joist_spacing_m;
  const joistCount = Math.max(2, Math.round(d / joistSpacing) + 1);
  const secondary_joists = joistCount * w;
  const bracing = 24.0;

  const nPortals = params.aframe.portal_count;
  const rafLen = rafterLengthM(params);
  const enclosed = params.envelope.enclosed_depth_m;
  const rafters = nPortals * 2 * rafLen;
  const tie_beams = nPortals * w;
  const ridge = enclosed + 0.4;
  const purlinRows = params.aframe.purlin_rows_per_side;
  const purlins = purlinRows * 2 * (enclosed + 0.4);

  const total = columns + main_beams + secondary_joists + bracing + rafters + tie_beams + ridge + purlins;
  return { columns, main_beams, secondary_joists, bracing, rafters, tie_beams, ridge, purlins, total };
}

function steelWeightKg(params, lengths) {
  return (
    lengths.columns          * profile(params, params.columns.profile).kg_per_m
    + lengths.main_beams       * profile(params, params.platform_structure.main_beam_profile).kg_per_m
    + lengths.secondary_joists * profile(params, params.platform_structure.secondary_joist_profile).kg_per_m
    + lengths.bracing          * profile(params, params.platform_structure.bracing_profile).kg_per_m
    + lengths.rafters          * profile(params, params.aframe.rafter_profile).kg_per_m
    + lengths.tie_beams        * profile(params, params.aframe.tie_beam_profile).kg_per_m
    + lengths.ridge            * profile(params, params.aframe.ridge_profile).kg_per_m
    + lengths.purlins          * profile(params, params.aframe.purlin_profile).kg_per_m
  );
}

function areas(params) {
  const enclosed = params.platform.width_m * params.envelope.enclosed_depth_m;
  const terrace = params.platform.width_m * params.envelope.terrace_depth_m;
  const platform = enclosed + terrace;
  const overhang = params.envelope.roof_overhang_m;
  const roof = 2 * rafterLengthM(params) * (params.envelope.enclosed_depth_m + 2 * overhang);
  return { enclosed_m2: enclosed, terrace_m2: terrace, platform_m2: platform, roof_m2: roof };
}

function costs(params, prices, areas) {
  // Anchor temporal — coincide con el cálculo de Python en M0.3.2.
  // En M0.3.3.1 se sustituye por una suma item-by-item ligada al BOM.
  const directCop = 218322300;

  const indirect = directCop * prices.overhead.indirect_pct;
  const contingency = directCop * prices.overhead.contingency_pct;
  const preIva = directCop + indirect + contingency;
  const iva = preIva * prices.overhead.iva_pct;
  const total = preIva + iva;
  const usd = total / prices.exchange_rate_usd_cop;
  return {
    direct_cop: directCop,
    indirect_cop: indirect,
    contingency_cop: contingency,
    pre_iva_cop: preIva,
    iva_cop: iva,
    total_cop: total,
    total_usd: usd,
    per_m2_enclosed_cop: total / areas.enclosed_m2,
  };
}

export function computeKpis(params, prices) {
  const lengths = steelLengths(params);
  const rawKg = steelWeightKg(params, lengths);
  const fabKg = rawKg * FAB_UPLIFT;
  const a = areas(params);
  const c = costs(params, prices, a);
  return {
    steel_lengths: lengths,
    steel_raw_kg: rawKg,
    steel_fabricated_kg: fabKg,
    anchors: columnCount(params) * params.columns.anchors_per_column,
    areas: a,
    cost: c,
    rafter_length_m: rafterLengthM(params),
  };
}

// ---------- Render helpers ----------

const fmtCOP = new Intl.NumberFormat("es-CO", { maximumFractionDigits: 0 });
const fmtUSD = new Intl.NumberFormat("en-US", { maximumFractionDigits: 0 });

function row(label, value, dirty = false) {
  return `<div class="kpi-row"><span class="kpi-label">${label}</span><span class="kpi-value ${dirty ? "dirty" : ""}">${value}</span></div>`;
}

function group(name) {
  return `<div class="kpi-group">${name}</div>`;
}

export function renderKpis(container, k, baseline) {
  const dirty = (curr, base) => Math.abs(curr - base) / Math.max(Math.abs(base), 1) > 0.001;

  const sl = k.steel_lengths;
  const slb = baseline.steel_lengths;
  const a = k.areas;
  const ab = baseline.areas;
  const c = k.cost;
  const cb = baseline.cost;

  container.innerHTML = [
    group("Geometría"),
    row("Plataforma",     `${a.platform_m2.toFixed(1)} m²`,  dirty(a.platform_m2, ab.platform_m2)),
    row("Cabaña cerrada", `${a.enclosed_m2.toFixed(1)} m²`,  dirty(a.enclosed_m2, ab.enclosed_m2)),
    row("Terraza",        `${a.terrace_m2.toFixed(1)} m²`,   dirty(a.terrace_m2,  ab.terrace_m2)),
    row("Cubierta",       `${a.roof_m2.toFixed(1)} m²`,      dirty(a.roof_m2,     ab.roof_m2)),
    row("Rafter",         `${k.rafter_length_m.toFixed(3)} m`, dirty(k.rafter_length_m, baseline.rafter_length_m)),

    group("Acero"),
    row("Longitud total", `${sl.total.toFixed(1)} m`, dirty(sl.total, slb.total)),
    row("Peso crudo",     `${k.steel_raw_kg.toFixed(0)} kg`, dirty(k.steel_raw_kg, baseline.steel_raw_kg)),
    row("Peso fabricado", `${k.steel_fabricated_kg.toFixed(0)} kg`, dirty(k.steel_fabricated_kg, baseline.steel_fabricated_kg)),
    row("≈ toneladas",    `${(k.steel_fabricated_kg / 1000).toFixed(2)} t`, dirty(k.steel_fabricated_kg, baseline.steel_fabricated_kg)),

    group("Anclajes"),
    row("Anclajes a roca", `${k.anchors} u`, dirty(k.anchors, baseline.anchors)),

    group("Costo (mid)"),
    row("Directo",      `$ ${fmtCOP.format(c.direct_cop)} COP`, dirty(c.direct_cop, cb.direct_cop)),
    row("Indirectos",   `$ ${fmtCOP.format(c.indirect_cop)} COP`, dirty(c.indirect_cop, cb.indirect_cop)),
    row("Contingencia", `$ ${fmtCOP.format(c.contingency_cop)} COP`, dirty(c.contingency_cop, cb.contingency_cop)),
    row("IVA 19%",      `$ ${fmtCOP.format(c.iva_cop)} COP`, dirty(c.iva_cop, cb.iva_cop)),
    row("TOTAL",        `$ ${fmtCOP.format(c.total_cop)} COP`, dirty(c.total_cop, cb.total_cop)),
    row("≈ USD",        `${fmtUSD.format(c.total_usd)} USD`, dirty(c.total_usd, cb.total_usd)),
    row("Por m² cabaña",`$ ${fmtCOP.format(c.per_m2_enclosed_cop)} COP`, dirty(c.per_m2_enclosed_cop, cb.per_m2_enclosed_cop)),
  ].join("");
}
